pkg/registry/generic/registry/store.go

```diff
@@ -234,6 +235,36 @@ func (e *Store) Create(ctx api.Context, obj runtime.Object) (runtime.Object, err
 	return out, nil
 }
 
+// shouldDelete checks if a Update is removing all the object's finalizers. If so,
+// it further checks if the object's DeletionGracePeriodSeconds is 0. If so, it
+// returns true.
+func (e *Store) shouldDelete(ctx api.Context, key string, obj, existing runtime.Object) bool {
+	if err := e.Storage.Get(ctx, key, existing, false); err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	// Note: we do not check doUnconditionalUpdate in this routine, because we
+	// expect in most cases the update of finalizers will be made by non-human,
+	// and we expect the resourceVersion will be properly set.
+
+	// validate the update.
+	if err := rest.BeforeUpdate(e.UpdateStrategy, ctx, obj, existing); err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	newMeta, err := api.ObjectMetaFor(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	oldMeta, err := api.ObjectMetaFor(existing)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false
+	}
+	return len(oldMeta.Finalizers) != 0 && len(newMeta.Finalizers) == 0 && newMeta.DeletionGracePeriodSeconds != nil && *newMeta.DeletionGracePeriodSeconds == 0
```

Comment:
> Also I'm not sure if deletion grace period seconds == 0 is the right thing to do

The BeforeUpdate() validates that the DeletionGracePeriods was not changed, so an update will only be translated into a delete if the DeletionGracePeriods of the old object is already 0, so it should be fine.